src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs

```diff
@@ -97,13 +97,51 @@ protected override void ExecuteCore()
 
                 var outputR2RImageRelativePath = file.GetMetadata(MetadataKeys.RelativePath);
                 var outputR2RImage = Path.Combine(OutputPath, outputR2RImageRelativePath);
+                bool crossgen2IsVersion5 = false;
+                if (ReadyToRunUseCrossgen2)
+                {
+                    string isVersion5 = Crossgen2Tool.GetMetadata(MetadataKeys.IsVersion5);
+                    crossgen2IsVersion5 = !string.IsNullOrEmpty(isVersion5) && bool.Parse(isVersion5);
+                }
+
+                string outputPDBImage = null;
+                string outputPDBImageRelativePath = null;
+                string crossgen1CreatePDBCommand = null;
+
+                if (EmitSymbols)
+                {
+                    if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && hasValidDiaSymReaderLib)
+                    {
+                        outputPDBImage = Path.ChangeExtension(outputR2RImage, "ni.pdb");
+                        outputPDBImageRelativePath = Path.ChangeExtension(outputR2RImageRelativePath, "ni.pdb");
+                        crossgen1CreatePDBCommand = $"/CreatePDB \"{Path.GetDirectoryName(outputPDBImage)}\"";
+                    }
+                    else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+                    {
+                        using (FileStream fs = new FileStream(file.ItemSpec, FileMode.Open, FileAccess.Read))
+                        {
+                            PEReader pereader = new PEReader(fs);
+                            MetadataReader mdReader = pereader.GetMetadataReader();
+                            Guid mvid = mdReader.GetGuid(mdReader.GetModuleDefinition().Mvid);
+
+                            outputPDBImage = Path.ChangeExtension(outputR2RImage, "ni.{" + mvid + "}.map");
+                            outputPDBImageRelativePath = Path.ChangeExtension(outputR2RImageRelativePath, "ni.{" + mvid + "}.map");
+                            crossgen1CreatePDBCommand = $"/CreatePerfMap \"{Path.GetDirectoryName(outputPDBImage)}\"";
+                        }
+                    }
+                }
 
                 if (!Crossgen2Composite)
                 {
                     // This TaskItem is the IL->R2R entry, for an input assembly that needs to be compiled into a R2R image. This will be used as
                     // an input to the ReadyToRunCompiler task
                     TaskItem r2rCompilationEntry = new TaskItem(file);
                     r2rCompilationEntry.SetMetadata(MetadataKeys.OutputR2RImage, outputR2RImage);
+                    if (outputPDBImage != null && ReadyToRunUseCrossgen2 && !crossgen2IsVersion5)
+                    {
+                        r2rCompilationEntry.SetMetadata(MetadataKeys.EmitSymbols, "true");
+                        r2rCompilationEntry.SetMetadata(MetadataKeys.OutputPDBImage, outputPDBImage);
+                    }
```

Comment:
We miss publishing the generated PDB file in Crossgen2 version 6 scenario.  Need to execute the code with `r2rSymbolsFileToPublish`.

Reply:
I see, thanks for the explanation, is it better now :-) ?

Reply:
That would not work for the composite mode, which generates a single PDB file.

Reply:
I have added a new section for publishing composite PDB / PerfMap, please take another look.